### PR TITLE
Rework the outline library and fix the outstanding issues with it

### PR DESCRIPTION
**Pull request by WardenPotato**
Its finally here, the rework of the outline library which started because of https://github.com/TTT-2/TTT2/issues/1411
Ive spent quite some time on this trying over 50 different approaches, running into various setbacks each time.
Some of them got solved: https://github.com/Facepunch/garrysmod-issues/issues/5744
And some didnt: https://github.com/Facepunch/garrysmod-requests/issues/2308

But this is what i consider the best thats possible in gmod today in respect to functionality and performance.

This rework includes(i may have forgotten some, its been a while):
- A switch to using the zbuffer instead of line of sight checking every single entity
- Outline thickness option
- Multiple render types for outlines
- More performant drawing
- Drawing not breaking when scopes are used in weapon addons
- Full documentation
- Fixed outlines being thinner on one side than the other

The comments in the code explain a lot so i hope those answer most questions.

### CODE REVIEW

File: lua/ttt2/libraries/outline.lua [hunk 1]
```diff
@@ -20,37 +25,172 @@ OUTLINE_MODE_BOTH = 0
...
-local storeTexture = render.GetScreenEffectTexture(0)
-local drawTexture = render.GetScreenEffectTexture(1)
+local copy_mat = Material("pp/copy")
 
 local ENTS = 1
 local COLOR = 2
 local MODE = 3
 
+--FIXME: You can decide if you want this to be a option for users, should be adressed before merge
+local enable_thin_line_workaround = false
```
Comment (TimGoll) [inline lua/ttt2/libraries/outline.lua:64]: we do have a performance section in the appearance menu. You could add it there if you like, enabled by default probably

File: lua/ttt2/libraries/outline.lua [hunk 2]
```diff
@@ -20,37 +25,172 @@ OUTLINE_MODE_BOTH = 0
...
-function outline.Add(ents, color, mode)
+function outline.Add(ents, color, mode, render_type, outline_thickness)
+    -- Make the default behaviour comply with older revisions
+    if not render_type then
+        render_type = OUTLINE_RENDERTYPE_AFTER_EF
+    end
+
+    if not outline_thickness then
+        outline_thickness = 1
+    end
```
Comment (TimGoll) [inline lua/ttt2/libraries/outline.lua:169]: is there a reason you didn't do `outline_thickness = outline_thickness or 1`, same for the other checks above

File: lua/ttt2/libraries/outline.lua [hunk 3]
```diff
@@ -20,37 +25,172 @@ OUTLINE_MODE_BOTH = 0
...
+
+    if not SetOutlineThickness(outline_thickness) then
+        return
+    end
+
+    -- Create list if it doesnt exist
+    InitializeCurrentListIfNeeded()
+
+    -- Get List and ListSize for this render type and outline thickness
+    local List, ListSize = GetCurrentList(), GetCurrentListSize()
```
Comment (TimGoll) [inline lua/ttt2/libraries/outline.lua:184]: variables in our lua style are always lower case - or is here a reason for this decision?

File: lua/ttt2/libraries/outline.lua [hunk 4]
```diff
@@ -75,137 +216,237 @@ function outline.RenderedEntity()
...
-        render.SetStencilReferenceValue(i)
-
-        surface.SetDrawColor(List[i][COLOR])
-        surface.DrawRect(0, 0, w, h)
-    end
-
-    cam.End2D()
-    -- end cam modification
+local MATERIAL_WHITE = Material("models/debug/debugwhite")
+local mat_antialias = GetConVar("mat_antialias")
```
Comment (TimGoll) [inline lua/ttt2/libraries/outline.lua:231]: is there a reason you chose two different naming styles? If you want to be in line with our style, it would be `materialDebugWhite` and `materialAntialias`

**Review by TimGoll — COMMENTED** [review]
I can't say much about the rendering stuff, but it does look really good to me, awesome work!

File: lua/ttt2/libraries/outline.lua (hunk 4 above)
Comment (WardenPotato) [reply lua/ttt2/libraries/outline.lua:231]: the naming i used here mostly reflects how its done by valve in the source engine, in there convars are snakecased similar to how they are in the console because it both communicates that its a convar and what that convar is without having to look it up. Midway this PR ive changed my personal preferences for naming aswell which can be seen

Comment (TimGoll) [reply lua/ttt2/libraries/outline.lua:231]: oh, the second one is no material, I'm blind :D

you can keep it that way or rename it to our style which is `cvEnableAntialias`

File: lua/ttt2/libraries/outline.lua (hunk 3 above)
Comment (WardenPotato) [reply lua/ttt2/libraries/outline.lua:184]: because of https://wiki.facepunch.com/gmod/list
though we could find another name

Comment (TimGoll) [reply lua/ttt2/libraries/outline.lua:184]: Ah, I see. I'd prefer a different name then. Something like `entsList`, `renderTargetList` or so maybe?

File: lua/ttt2/libraries/outline.lua (hunk 1 above)
Comment (WardenPotato) [reply lua/ttt2/libraries/outline.lua:64]: It really hinges on whether you find the appearance of them being slighty thicker fine, its super subjective but it does look better than the thin line scenarios where it can get cutoff

File: lua/ttt2/libraries/outline.lua (hunk 2 above)
Comment (TimGoll) [reply lua/ttt2/libraries/outline.lua:169]: fair, I don't care that much. I just assumed it had a performance impact and was chosen for that reason

File: lua/ttt2/libraries/outline.lua (hunk 1 above)
Comment (TimGoll) [reply lua/ttt2/libraries/outline.lua:64]: you know what, if the performance is mostly similar, chose whatever you believe looks best and set it as default

File: lua/ttt2/libraries/outline.lua (hunk 2 above)
Comment (WardenPotato) [reply lua/ttt2/libraries/outline.lua:169]: Screenshot_20240922_232117
Screenshot_20240922_232107
Bytecode seems to be the exact same so either will do fine, just a style preference at that point

Comment (TimGoll) [reply lua/ttt2/libraries/outline.lua:169]: Keep it as is, that's fine :)

File: lua/ttt2/libraries/outline.lua (hunk 3 above)
Comment (WardenPotato) [reply lua/ttt2/libraries/outline.lua:184]: Ah i remember why i did that, theres also a global Lists and ListSize, i used shadowing locals there to make development easier keeping it portable between different parts of the module, but it does break the no uppercase for non globals rule, can adjust that if you still want it

File: lua/ttt2/libraries/outline.lua (hunk 1 above)
Comment (WardenPotato) [reply lua/ttt2/libraries/outline.lua:64]: no perf difference, it just adjusts the offset by 1 pixel, theres 1 edge case where this looks worse and thats if you have a 1 pixel thick line of geometry on a model being outlined but thats exceptionally rare and even then you'd have to zoom in to notice it, the problem with lines being too thin is a lot more common and noticeable, i'll leave what looks best to you guys in this case because my opinion keeps flipping dependant on which ive been looking at longer so im kinda neutral on it

Comment (TimGoll) [reply lua/ttt2/libraries/outline.lua:64]: to me your discription sounds like it should be enabled

**Comment by TimGoll** [comment]
Okay, I tried tackling some of the warnings here and there is one thing I'd like you to change: `List`, `Lists`, `ListSize`, `ListsSize` are way too similar and I have no clue what they are about. Maybe make the name a bit more descriptive.
I don't know what the performance inpact would be, but it would also make sense to put it into the `outline` scope imho.

Once you get that done and the workflow errors are fixed, I can merge this PR

**Comment by WardenPotato** [comment]
Performance impact should be zero because in the worst case i can localise the global anyways, ill get to changing the names then

**Comment by nike4613** [comment]
What still needs to be done on this for it to be mergeable?

**Comment by WardenPotato** [comment]
> What still needs to be done on this for it to be mergeable?

The workflow errors need fixing

There was a suggestion to make the optional thin line fix a toggleable option in the settings menu

and TimGolls comment about variable naming

**Comment by saibotk** [comment]
Jeah we can take over this and fix the rest.
Just a bit slow atm due to vacancies etc. @Histalek and I will try to get this in soon

File: lua/ttt2/libraries/outline.lua (hunk 1 above)
Comment (Histalek) [reply lua/ttt2/libraries/outline.lua]: i'm with tim on this one

enabled unconditionally in 4fa8e774e8d40c943d204ce3aa653979cfe3bfe4

File: lua/ttt2/libraries/outline.lua (hunk 3 above)
Comment (Histalek) [reply lua/ttt2/libraries/outline.lua:174]: the inconsistency is already there with how gmod itself handles things ... besides this was fine before so i don't see a reason to change this now

**Comment by Histalek** [comment]
i'm struggling with adding a proper changelog to this, otherwise this should now be good to go IMO

**Comment by TimGoll** [comment]
> i'm struggling with adding a proper changelog to this, otherwise this should now be good to go IMO

@WardenPotato could you summarize your work as a changelog? Otherwise I'd just mention these things:

Change: Outline now support different thicknesses
Fixed: Outline Interactions with certain weapon's scopes
Fixed: Outlines not rendering uniformly

**Comment by WardenPotato** [comment]
> > i'm struggling with adding a proper changelog to this, otherwise this should now be good to go IMO
> 
> @WardenPotato could you summarize your work as a changelog? Otherwise I'd just mention these things:
> 
> Change: Outline now support different thicknesses Fixed: Outline Interactions with certain weapon's scopes Fixed: Outlines not rendering uniformly

whats in the pull description is pretty much what has been done, i dont have more to add to that

**Review by TimGoll — APPROVED** [review]

**Review by saibotk — APPROVED** [review]
